Fix [LB-13] | Export Listens | JSON

pinkeshbadjatiya opened the pull request:
Fix for <b><a href="http://tickets.musicbrainz.org/browse/LB-13">LB-13</a></b>
Listen export to json file.
![Screenshot](http://i.imgur.com/J1GfWsa.png)
<hr/>
![Screenshot](http://i.imgur.com/m3iXHKV.png)

gentlecat commented on listenstore/listenstore/listenstore.py:
```diff
@@ -218,10 +218,18 @@ def fetch_listens_for_range(self, uid, date_range, from_id, to_id, order, limit=
-                range_keys(len(date_range)) + \
-                """ AND id > %(from_id)s AND id < %(to_id)s
-                   ORDER BY id """ + ORDER_TEXT[order] + """ LIMIT %(limit)s"""
+        query = "" 
```
This is not needed.

pinkeshbadjatiya replied:
done.

gentlecat commented on webserver/templates/user/export.html:
```diff
@@ -0,0 +1,31 @@
+     </p>
+
+     <p>
+		Now you can export you listens to a file.
```
Don't mix tabs and spaces.

gentlecat commented on webserver/templates/user/export.html:
```diff
@@ -0,0 +1,31 @@
+         </form>
+     </div>
+
+	 {% endif %}
```
Check indentation.

gentlecat commented on webserver/views/user.py:
```diff
@@ -98,6 +103,31 @@ def import_data():
+        return render_template("user/export.html", user=current_user)
+
+
+
```
Two empty lines between functions. http://pep8.org/#blank-lines

gentlecat commented on webserver/views/user.py:
```diff
@@ -98,6 +103,31 @@ def import_data():
+        cassandra = webserver.create_cassandra()
+        user_id = current_user.musicbrainz_id
+        
+        ii = 1
```
What is this?

gentlecat replied:
Also, you can use https://docs.python.org/2/library/functions.html#enumerate.

pinkeshbadjatiya replied:
Surely a useful suggestion. Thanks.

gentlecat commented on webserver/views/user.py:
```diff
@@ -98,6 +103,31 @@ def import_data():
+@user_bp.route("/export")
+@login_required
+def export_data():
+    filename = request.args.get("filename")
```
Why do we need a file name?

pinkeshbadjatiya replied:
We can actually just export the data, but providing user an option to give file name won't be a bad idea.

gentlecat replied:
Can't user choose a file name when saving a file? Browsers have this feature.

pinkeshbadjatiya replied:
At least for firefox, i just have a "Save As" and "Open With" option. Mostly depends on user's configuration, i think.

gentlecat replied:
I still don't think we should have this option.

pinkeshbadjatiya replied:
ok. Is something like, <code>gentlecat_LB.json</code> ok? Any other suggestion?

gentlecat replied:
Sure. I'd also make it completely lowercase.

pinkeshbadjatiya replied:
ok

pinkeshbadjatiya replied:
should i make it something like .<code>../export?type=json</code>

gentlecat commented on webserver/views/user.py:
```diff
@@ -98,6 +103,31 @@ def import_data():
+        filename = secure_filename(filename)
+        output = {}
+        cassandra = webserver.create_cassandra()
+        user_id = current_user.musicbrainz_id
```
`current_user.musicbrainz_id` can be passed directly to `fetch_listens` function call. No need to put it in a variable.

pinkeshbadjatiya replied:
will make the appropriate changes.

gentlecat commented on listenstore/listenstore/listenstore.py:
```diff
@@ -218,10 +218,17 @@ def fetch_listens_for_range(self, uid, date_range, from_id, to_id, order, limit=
+                    """ AND id > %(from_id)s AND id < %(to_id)s
+                    ORDER BY id """ + ORDER_TEXT[order] + """ LIMIT %(limit)s"""
+        else:
+            query =     """SELECT * FROM listens WHERE uid = %(uid)s AND """ + \
```
Spacing is a bit off there.

pinkeshbadjatiya replied:
done

pinkeshbadjatiya commented:
Fixed all the issues mentioned above. Ready for merge. :)

pinkeshbadjatiya commented:
Fixed all the issues mentioned above. Ready for merge. :)

pinkeshbadjatiya commented:
Fixed all the issues mentioned above. Ready for merge. :)

gentlecat commented:
```
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1836, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1820, in wsgi_app
    response = self.make_response(self.handle_exception(e))
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1403, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/usr/local/lib/python2.7/dist-packages/flask_login.py", line 758, in decorated_view
    return func(*args, **kwargs)
  File "/home/vagrant/listenbrainz/webserver/views/user.py", line 118, in export_data
    is_json=True), start=1))
  File "/home/vagrant/listenbrainz/listenstore/listenstore/listenstore.py", line 203, in fetch_listens
    order, current_limit, is_json):
  File "/home/vagrant/listenbrainz/listenstore/listenstore/listenstore.py", line 251, in fetch_listens_for_range
    results = self.execute(query, params)
  File "/home/vagrant/listenbrainz/listenstore/listenstore/listenstore.py", line 121, in execute
    return self.session.execute(query.encode(), params)
  File "build/bdist.linux-x86_64/egg/cassandra/cluster.py", line 1295, in execute
    result = future.result(timeout)
  File "build/bdist.linux-x86_64/egg/cassandra/cluster.py", line 2799, in result
    raise self._final_exception
SyntaxException: <ErrorMessage code=2000 [Syntax error in CQL query] message="line 1:12 no viable alternative at input '*' (SELECT [JSON] *...)">
```

gentlecat commented on webserver/templates/user/export.html:
```diff
@@ -0,0 +1,23 @@
+		</p>
+		<h3>
+			<strong> Export </strong>
+		</h3>
```
This title is redundant.

gentlecat commented on webserver/templates/user/export.html:
```diff
@@ -0,0 +1,23 @@
+		</h3>
+		<form action="{{ url_for("user.export_data") }}" method="GET">
+			<input type="hidden" name="type" value="json">
+			<input type="submit" value="Download Listens as json">
```
I would just set value to `Download` and describe that dump is in JSON format.

gentlecat commented on webserver/templates/user/export.html:
```diff
@@ -0,0 +1,23 @@
+		</p>
+		<p>
+				Now you can export your listens to a file.
+				Just choose one of the export formats and let ListenBrainz do the hard work of figuring out what to do.
```
There is only one export format.

pinkeshbadjatiya replied:
We may be adding more formats in future, so i wrote a generic line. :)
Anyways, will change asap.

pinkeshbadjatiya commented:
regarding the error. thats wierd. For me its working fine. Also it seems to be error in the CQL query syntax. I have updated the commit. Can you recheck and tell if it still exists?

alastair commented:
Please check the version of cassandra you are using. It looks like `SELECT JSON` was introduced in 2.2, and we are using 2.1.9 in ListenBrainz. This is why it is important that you use the existing development infrastructure which we provide, or that you replicate it exactly if you install it yourself.

pinkeshbadjatiya commented:
Fixed the JSON support for Cassandra 2.1.9

pinkeshbadjatiya commented:
Ready for merge. :)

gentlecat commented:
Export seems to be working now, but I'm not sure about structure of the dump.

```JSON
{
  "0": {
    "album_msid": "None",
    "timestamp": 1457811780,
    "artist_msid": "f8d51abc-c69e-4108-97f6-0bd008c9f749",
    "artist_name": "Perturbator",
    "recording_msid": "ef75aca2-7b85-4ee4-bdf8-11beead5e58c",
    "track_name": "Miami Disco",
    "additional_info": {
      "artist_msid": "f8d51abc-c69e-4108-97f6-0bd008c9f749",
      "spotify_id": "http:\/\/open.spotify.com\/track\/5QyvmZYiOuK0saTv6J77Yg",
      "release_msid": null
    }
  },
  "1": {
    "album_msid": "None",
    "timestamp": 1457811300,
    "artist_msid": "c00abe3d-48fc-4eaa-94aa-e15d7f7e102c",
    "artist_name": "Moderat",
    "recording_msid": "58417c9c-d49e-4112-817e-243bf9584131",
    "track_name": "Reminder",
    "additional_info": {
      "artist_msid": "c00abe3d-48fc-4eaa-94aa-e15d7f7e102c",
      "release_msid": null
    }
  },
  ...
}
```

Why do we need these indexes (keys) instead of just having a list of listens?

gentlecat commented on listenstore/listenstore/listenstore.py:
```diff
@@ -242,6 +241,7 @@ def fetch_listens_for_range(self, uid, date_range, from_id, to_id, order, limit=
 
             self.log.debug("Fetching up to %s rows for date_range: %s, uid: %s", this_limit, date_range, uid)
             results = self.execute(query, params)
+            output = []
```
This is unused. I would also suggest to remove other unrelated changes in this file or move them to a separate commit/pull request.

gentlecat commented on .gitignore:
```diff
@@ -4,3 +4,4 @@
 *.pyc
 config.py
 *.log
+.*.swp
```
Where are these files coming from?

pinkeshbadjatiya replied:
When opening the files in vim, it creates .swp files when the files are open.
So if someone commits while a file is open, then we have unnecessary files in the commit.

gentlecat replied:
I would suggest to use a [global .gitignore](https://help.github.com/articles/ignoring-files/#create-a-global-gitignore) file for things that are not directly related to the project.

gentlecat commented on webserver/templates/user/export.html:
```diff
@@ -0,0 +1,23 @@
+{% extends 'base.html' %}
+{% block title %} Export for "{{ user.musicbrainz_id }}" - ListenBrainz {% endblock %}
```
You don't need to put spaces after opening and before closing tag.

gentlecat commented on webserver/templates/user/export.html:
```diff
@@ -0,0 +1,23 @@
+{% block title %} Export for "{{ user.musicbrainz_id }}" - ListenBrainz {% endblock %}
+
+{% block content %}
+	{% if current_user.is_authenticated() and current_user == user %}
```
Use indentation consistently with other files in the project. For HTML templates it's 2 spaces, not tabs.

gentlecat commented on webserver/views/user.py:
```diff
@@ -19,7 +21,8 @@ def lastfmscraper(user_id):
         "user/scraper.js",
-        base_url=url_for("api_v1.submit_listen", user_id=user_id, _external=True),
+        base_url=url_for("api_v1.submit_listen",
+            user_id=user_id, _external=True),
```
Again, this change is unnecessary. Check all other changes, I see plenty of similar ones that don't server any purpose.

pinkeshbadjatiya replied:
Should i remove the `docstrings` as well?

gentlecat commented on webserver/views/user.py:
```diff
@@ -6,6 +6,8 @@
 from datetime import datetime
 import webserver
 import db.user
+import ujson
```
Why did you decide to use `ujson` package instead of `json`? We encountered some problems with accuracy when converting some values in AcousticBrainz using `ujson`. Do you know if this is relevant here?

pinkeshbadjatiya replied:
I was unaware of that.
I actually saw it being used somewhere in the code, and when i read about it i thought it would be good to use it.

> UltraJSON is an ultra fast JSON encoder and decoder written in pure C wit...

I'll check and convert all instances to use `json` instead.

gentlecat replied:
I'm not saying that you need to use `json` package instead. Just asking for a reason.

pinkeshbadjatiya replied:
Can you explain the problem faced with using ujson? Or can you point to the ticket?  It might be useful in future. :)

alastair replied:
`ujson` gave us problems after about 10 decimal places on floating point numbers. We started using it in ListenBrainz because we were serialising and deserialising a lot of data and it's about 3x faster than regular `json`. For this task it's OK to continue using `ujson`

gentlecat replied:
:ok_hand:

gentlecat commented on webserver/views/user.py:
```diff
@@ -98,7 +104,35 @@ def import_data():
+def export_data():
+    """ Exporting the data to various formats, like, json """
+    filetype = request.args.get("type")
+    if filetype == "json":
```
I don't think this is a good way to determine if we need to return data or render the interface. How about sending POST request with format of the dump?

I'm also not sure if specifying the format is necessary. Are there any concrete plans to add alternatives in near future? If not, this seems useless.

pinkeshbadjatiya replied:
As of now i don't think there are any concrete plans to add new export format. Also, actually we just need a single format to export.
Adding POST request.

pinkeshbadjatiya commented:
> Why do we need these indexes (keys) instead of just having a list of listens?

I wanted the export to be in `valid JSON` so that if we want to reimport, then we just need to jsonize it and we have all the data in correct format. I earlier tried to create a minimilistic file with removed formatting, but this looked better and made sense.

pinkeshbadjatiya commented:
Changed export to `POST` request.
I had a few question,
- *pylint* shows a lot of `line-too-long` errors. Earlier many changes were related to it. Should i keep them bcoz i just reverted as you requested.
- I added a few `docstrings` as well. I prefer to keep them.

Any suggestions?

gentlecat commented:
> *pylint* shows a lot of `line-too-long` errors. Earlier many changes were related to it. Should i keep them bcoz i just reverted as you requested

Don't worry about long lines. Just make sure it's readable.

> I added a few `docstrings` as well. I prefer to keep them.

Sure, docstrings are useful. Keep them. You can also improve formatting a bit (https://www.python.org/dev/peps/pep-0257/#one-line-docstrings).

alastair commented:
I've attached an example of a last.fm dump in json format. It is also presented as a list. I agree with @gentlecat, I think it's better to have a list than a dictionary.
[scrobbles-2010-10.json.txt](https://github.com/metabrainz/listenbrainz-server/files/184400/scrobbles-2010-10.json.txt)

gentlecat commented on webserver/views/user.py:
```diff
@@ -98,7 +102,34 @@ def import_data():
                            lastfm_username=lastfm_username)
 
 
+@user_bp.route("/export", methods=["GET","POST"])
```
Space after comma.

gentlecat commented on webserver/views/user.py:
```diff
@@ -98,7 +102,34 @@ def import_data():
+
+        # Fetch output and convert it into dict with keys as indexes
+        output = {}
+        for index,obj in enumerate(cassandra.fetch_listens(current_user.musicbrainz_id)):
```
Space after comma.

pinkeshbadjatiya commented:
Current Format -
```
[{
	"album_msid": "None",
	"timestamp": 1458836372,
	"artist_msid": "487f6315-8d02-42ce-9159-1ec7976499da",
	"artist_name": "Woolfy",
	"recording_msid": "c59692f3-f54e-4ab4-ac13-69ecc82191f2",
	"track_name": "Tangiers",
	"additional_info": {
		"artist_msid": "487f6315-8d02-42ce-9159-1ec7976499da",
		"release_msid": null
	}
}, {
	"album_msid": "None",
	"timestamp": 1458836020,
	"artist_msid": "f602
       ....
       ....
}]

```

alastair commented:
This is looking really good, thanks! Some comments:
* `"album_msid": "None"` - we shouldn't have `None` in quotes here. None is a python thing. In json this should be `null`, like in `"release_msid": null`
* The filename should have the current date in it, in yyyy-mm-dd format

Some other things which we also need to address:
* We haven't been pushing for it, but we really need to start including tests. I would like to see this code tested, but if you prefer you can do it in a separate PR (just this once, all subsequent PRs must have tested code). If you have questions about how do good testing, ask us in IRC.
* We should make a settings page. We can link to this export (and import too) from settings, instead of continuing to make the menu bigger. You can include this page on this PR or in the import PR, I don't mind.

pinkeshbadjatiya commented:
- Made appropriate changes  

---
- I would love to include tests. I believe tests can be done on a separate PR, this would make things easier for me, and would keep things simpler. :)
- **Yes!!**  - I read about this idea on one of the tickets on jira. It would be better to provide users with a settings page. I think, we can wait a bit to see what is the outcome of `export-PR` and `import-PR`. Once they get merged/closed, then we can start working on the settings page.

alastair commented:
Ready to merge. I will merge it and make any cleanups if needed

mineo commented on webserver/views/user.py:
```diff
@@ -98,7 +102,35 @@ def import_data():
+
+        # Fetch output and convert it into dict with keys as indexes
+        output = []
+        for index, obj in enumerate(cassandra.fetch_listens(current_user.musicbrainz_id)):
```
Just a general question about this approach: does this work reliably for users with a large (>100k entries) listen history? I've retrieved all my listens (~130000, a modified version of the JSON data is ~34MB) doing 1 request per second and with that approach I saw intermittent streaks of HTTP 502 (or 503 or 504, the timeout ones in any case) status codes.

mayhem commented on webserver/templates/user/export.html:
```diff
@@ -0,0 +1,20 @@
+    </p>
+    <p>
+        Now you can export your listens to a file. 
+        Just click on the download and let ListenBrainz do the hard work of figuring out what to do.
```
Please update lines 11 & 12 to say:

Export your listen history in JSON format and download it. Click Download to proceed.

mayhem commented:
I've test merged this into my docker-without-kafka branch and it works fine. Please update the text as requested and then I'l merge this after the docker branch is merged.
